### PR TITLE
ZMQ sockets (zsock_t) do not work in child process after fork #2038

### PR DESCRIPTION
**Pull request by santupari**
We are observing an issue where in sockets created in child process do not work after forking.
Below is the sequence of code and I have also attached test code that reproduces issue and as well as working code:

Our design has a server which listens to multiple clients and responds
Clients uses wrapper API (example : test_tx_rx()) that exchanges messages with server
the issue observed when a client calls test_tx_rx() from parent and then fork and if you call test_tx_rx() from child it fails to communicate with server
The fix is to call zsys_shutdown() from child to clean up all initialization done by parent and assume that ZSYS is already initialized. However, as part of this fix we also noticed that there is a mutex that should be re-initialized to avoid deadlock. Hence, we are also attaching a patch that defines new API for re-initializing the mutex.

Please let us know if the patch is valid or how to avoid deadlock in case if the parents threads are accessing the mutex at the same time when child is clearing mutex in zsys_shutdown().

I am also attaching test code that can be used to reproduce the issue. This issue is no longer observed after the patch is applied.

Thanks,
Santosh Bidaralli
[zsys_patch_wtih_test_code.zip](https://github.com/zeromq/czmq/files/5247679/zsys_patch_wtih_test_code.zip)

### CODE REVIEW

**Comment by bluca** [comment]
the windows build are broken

File: src/zsys.c [hunk 1]
```diff
@@ -150,13 +153,46 @@ s_zsys_vprintf_hint (int hint, const char *format, va_list argptr);
 //  Not threadsafe, so call only from main thread. Safe to call multiple
 //  times. Returns global CZMQ context.
 
+// mutex for pthread_once to run the init function only once in a process
+static pthread_once_t init_all_mutex_var = PTHREAD_ONCE_INIT;
+
+// handler to initialize mutexes one time in multi threaded env
+void zsys_initialize_mutex() {
```
Comment (bluca) [inline src/zsys.c:160]: all these new functions should be static

Comment (santupari) [reply src/zsys.c:160]: i will send a new patch after changing these functions to static.

**Comment by bluca** [comment]
the build still fails on windows, it needs to be fixed before this can be merged

**Comment by santupari** [comment]
Thanks for the response. I will take a look at get back with a fix at the earliest.

**Comment by bluca** [comment]
src/zsys.c is getting the executable bit set - likely a problem with your editor. Please remove it.

**Comment by santupari** [comment]
I am sorry for the inconvenience. I have pushed the new file after removing executable bit.

**Pull request merged**